src/mds/Locker.cc

```diff
@@ -2991,6 +2992,10 @@ void Locker::_update_cap_fields(CInode *in, int dirty, MClientCaps *m, inode_t *
 	      << " for " << *in << dendl;
       pi->mtime = mtime;
     }
+    if (((dirty & CEPH_CAP_FILE_WR) && change_attr > pi->change_attr) ||
+	((dirty & CEPH_CAP_FILE_EXCL) && change_attr != pi->change_attr)) {
+      pi->change_attr = change_attr;
```

Comment:
Hmm, actually now that I think about it, I don't think I quite have the right semantics for the change_attr when we have multiple clients. I put the change_attr under CEPH_CAP_FILE_EXCL capability, but I don't think that's quite sufficient.

Suppose we have 2 clients -- one with CEPH_CAP_FILE_EXCL and one with CEPH_CAP_AUTH_EXCL. The AUTH_EXCL client does (e.g.) a mode change, while the other client does a single write. Both bump the change_attr by 1.

Next, the server ends up recalling caps from both clients, and the FILE_EXCL client then checks the change_attr again. It hasn't changed though the mode has...

Actually, now that I think about it, that's probably OK since you're not allowed to infer anything about the value of the change_attr. Applications must treat this as an opaque token that can only be checked for equivalence.

Between the point where the client fetched the mode for its original check and where it does the write, it's ok for the mode to change. All the client knows is that after the write, the change_attr was bumped, and so any cached data and metadata is invalid at that point.